File: Python/Day20.py
"""
Santa Claus 🎅 está revisando la lista de regalos que debe entregar esta Navidad. Sin embargo, algunos regalos faltan, otros están duplicados, y algunos tienen cantidades incorrectas. Necesita tu ayuda para resolver el problema.

Recibirás dos arrays:

received: Lista con los regalos que Santa tiene actualmente.
expected: Lista con los regalos que Santa debería tener.
Tu tarea es escribir una función que, dado received y expected, devuelva un objeto con dos propiedades:

missing: Un objeto donde las claves son los nombres de los regalos faltantes y los valores son las cantidades que faltan.
extra: Un objeto donde las claves son los nombres de los regalos extra o duplicados y los valores son las cantidades que sobran.
Ten en cuenta que:

Los regalos pueden repetirse en ambas listas.
Las listas de regalos están desordenadas.
Si no hay regalos que falten o sobren, las propiedades correspondientes (missing o extra) deben ser objetos vacíos.
"""

def fix_gift_list(received: list[str], expected: list[str]) -> dict[str, int]:
    result = {
        "missing": {},
        "extra": {}
    }

    for i in range(len(expected)):
        gift = expected[i]
       

        if gift in received:
            index = received.index(gift)
            received[index] = None
            expected[i] = None
        else:
            if not gift in result["missing"]:
                result["missing"][gift] = 1
            else:           
                result["missing"][gift] +=1
            

    for gift in received:
        if gift == None : continue
       
        if not gift in expected:            
            if not gift in result["extra"]:
                result["extra"][gift] = 1
            else:
                result["extra"][gift]+=1

    return result

print(fix_gift_list(
  ['puzzle', 'car', 'doll', 'car'], ['car', 'puzzle', 'doll', 'ball']
))

print(fix_gift_list(
   ['book', 'train', 'kite', 'train'],
  ['train', 'book', 'kite', 'ball', 'kite']
))

File: JS/Day20.js
/*
Santa Claus 🎅 está revisando la lista de regalos que debe entregar esta Navidad. Sin embargo, algunos regalos faltan, otros están duplicados, y algunos tienen cantidades incorrectas. Necesita tu ayuda para resolver el problema.

Recibirás dos arrays:

received: Lista con los regalos que Santa tiene actualmente.
expected: Lista con los regalos que Santa debería tener.
Tu tarea es escribir una función que, dado received y expected, devuelva un objeto con dos propiedades:

missing: Un objeto donde las claves son los nombres de los regalos faltantes y los valores son las cantidades que faltan.
extra: Un objeto donde las claves son los nombres de los regalos extra o duplicados y los valores son las cantidades que sobran.
Ten en cuenta que:

Los regalos pueden repetirse en ambas listas.
Las listas de regalos están desordenadas.
Si no hay regalos que falten o sobren, las propiedades correspondientes (missing o extra) deben ser objetos vacíos.
*/
/**
 * @typedef {Record<string, number>} GiftsCount
 */

/**
 * @typedef {{ missing: GiftsCount, extra: GiftsCount }} Result
 */

/**
 * @param {string[]} received
 * @param {string[]} expected
 * @returns {Result}
 */
function fixGiftList(received, expected) {
    let result = {
        missing: {},
        extra: {}
    };

    for(let i=0;i<expected.length;i++){
        let gift = expected[i];    
        let index = received.indexOf(gift);

        if(index !== -1) {
            received[index] = null;
            expected[i] = null;
        } else{
            if(result.missing[gift] === undefined){
                result.missing[gift] = 1;
            }
            else{
                result.missing[gift]++;
            }
        }
    }

    for(let gift of received){
        if(gift === null) continue;
        let index = expected.indexOf(gift);
        if(index === -1) {
            if(result.extra[gift] === undefined){
                result.extra[gift] = 1;
            }
            else{
                result.extra[gift]++;
            }
        }
    }

    return result;
  }

console.log( fixGiftList(['puzzle', 'car', 'doll', 'car'], ['car', 'puzzle', 'doll', 'ball']));
// Devuelve:
// {
//   missing: { ball: 1 },
//   extra: { car: 1 }
// }

console.log(fixGiftList(
  ['book', 'train', 'kite', 'train'],
  ['train', 'book', 'kite', 'ball', 'kite']
));
// Devuelve:
// {
//   missing: { ball: 1, kite: 1 },
//   extra: { train: 1 }
// }

console.log(fixGiftList(
  ['bear', 'bear', 'car'],
  ['bear', 'car', 'puzzle', 'bear', 'car', 'car']
));
// Devuelve:
// {
//   missing: { puzzle: 1, car: 2 },
//   extra: {}
// }

console.log(fixGiftList(['bear', 'bear', 'car'], ['car', 'bear', 'bear']));
// Devuelve:
// {
//   missing: {},
//   extra: {}
// }